### PR TITLE
Multi-Variant Architecture

### PR DESCRIPTION
**Pull request by mriedmann**
To make it possible to separate the nginx and PHP part for production setups we are forced to change the architecture a bit. This should not break existing setups but will provide us with more possibilities.

### CODE REVIEW

File: docker-compose.prod.yml [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    ports:
+      - "8080:80"
+    healthcheck:
+      test: ["CMD", "curl", "-f", "http://localhost/ping"]
+      interval: 30s
+      timeout: 10s
+      retries: 3
+    volumes:
+      - "uploads:/var/www/localhost/htdocs/uploads"
+      - "assets:/var/www/localhost/htdocs/assets"
```
Comment (jvies) [inline docker-compose.prod.yml:23]: add the following volumes:
"modules:/var/www/localhost/htdocs/protected/modules"
"themes:/var/www/localhost/htdocs/themes"

File: docker-compose.prod.yml [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+      HUMHUB_CACHE_EXPIRE_TIME: 3600
+      HUMHUB_CACHE_CLASS: yii\redis\Cache
+      HUMHUB_QUEUE_CLASS: humhub\modules\queue\driver\Redis
+      HUMHUB_REDIS_HOSTNAME: redis
+      HUMHUB_REDIS_PORT: 6379
+      HUMHUB_REDIS_PASSWORD: redis_password
+    volumes:
+      - "uploads:/var/www/localhost/htdocs/uploads"
+      - "assets:/var/www/localhost/htdocs/assets"
+#     - "${HUMHUB_VOLUME_CONFIG}:/var/www/localhost/htdocs/protected/config"
```
Comment (jvies) [inline docker-compose.prod.yml:59]: add the following volumes:
"modules:/var/www/localhost/htdocs/protected/modules"
"themes:/var/www/localhost/htdocs/themes"

File: docker-compose.prod.yml [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    healthcheck:
+      test: ["CMD", "php-fpm-healthcheck"]
+      interval: 30s
+      timeout: 10s
+      retries: 3
+  db:
+    image: mariadb
```
Comment (jvies) [inline docker-compose.prod.yml:66]: use a tagged version of mariadb with major.minor tag as mariadb:10.4 so it does not pull the latest release

**Comment by mriedmann** [comment]
I am quite satisfied with this PR for now. 

@madmath03 Are you up for a short review? Is that what you had in mind with #124?

It also paves the way for releasing images using GitHub Actions rather than Docker Hub Auto-Build. Using this method we can move the whole repo to humhub/docker, publishing images to mriedmann/humhub and humhub/humhub at the same time.

**Review by madmath03 — COMMENTED** [review]
Hi @mriedmann 

I did not read everything in details, but overall it does seem to allow do what I aimed with #124 : be able to split Nginx and PHP FPM in 2 specific containers :+1:

I do have a few questions though (probably due not reading thoroughly enough):
1. what about "_legacy_" standalone behavior ? is it still available ?
2. are their going to be 3 docker tags (1 for nginx, 1 for PHP only, 1 for allinone) for each humhub versions ?

**Comment by mriedmann** [comment]
Yes and yes :)

The Allinone Tag should behave exactly like the current version. I sm planning to publish 2 additional tags: nginx and phponly as variants to make it possible to split the static content from php. The problem here is, that humhub itself has to provide some content (/uploads, /templates, /modules,/assets). Therefore, it is nessacery to allow php write access to the volumes of the nginx container, making this whole idea a bit less usefull. Nevertheless it makes it easier to use a shared nginx layer (not provided by this project). 
What was your initial intention for this change? Is it met with this PR?

**Comment by jvies** [comment]
Hi,

Any idea when the PR will be merged ?

I'd like to update my container.

I've seen Humhub released the version 1.6.4. Maybe you can update the PR with this new tag.

Great job on this image by the way.

best regards,
Jérémy

**Comment by mriedmann** [comment]
I am quite busy lately, will try to get this one merged end released this weekend. We can do the 1.6.4 update right after that.

**Comment by madmath03** [comment]
> What was your initial intention for this change? Is it met with this PR?

Yes @mriedmann , this was the idea of #124 but your way definitely seems better :+1:

**Pull request merged**